urlparse4/cgurl.pyx

```diff
@@ -333,13 +392,157 @@ def urljoin(base, url, allow_fragments=True):
     This function intends to replace urljoin from urllib,
     which uses Resolve function from class GURL of GURL chromium
     """
+    str_input = isinstance(base, str)
+    if isinstance(url, str) != str_input:
+        raise TypeError("Cannot mix str and non-str arguments")
+
     decode = not (isinstance(base, bytes) and isinstance(url, bytes))
     if allow_fragments and base:
         base, url = unicode_handling(base), unicode_handling(url)
+        """
+        this part needs to be profiled to see if creating another GURL instance
+        here takes more time than expected?
+        """
+        if not GURL(base).is_valid():
+            fallback = stdlib_urljoin(base, url, allow_fragments=allow_fragments)
+            if decode:
+                return fallback.decode('utf-8')
+            return fallback
         joined_url = GURL(base).Resolve(url).spec()
 
         if decode:
             return joined_url.decode('utf-8')
         return joined_url
 
     return stdlib_urljoin(base, url, allow_fragments=allow_fragments)
+
+def _safe_ParseResult(parts, encoding='utf8', path_encoding='utf8'):
+    # IDNA encoding can fail for too long labels (>63 characters)
+    # or missing labels (e.g. http://.example.com)
+    try:
+        netloc = parts.netloc.encode('idna')
+    except UnicodeError:
+        netloc = parts.netloc
+
+    return (
+        to_native_str(parts.scheme),
+        to_native_str(netloc),
+
+        # default encoding for path component SHOULD be UTF-8
+        quote(to_bytes(parts.path, path_encoding), _safe_chars),
+        quote(to_bytes(parts.params, path_encoding), _safe_chars),
+
+        # encoding of query and fragment follows page encoding
+        # or form-charset (if known and passed)
+        quote(to_bytes(parts.query, encoding), _safe_chars),
+        quote(to_bytes(parts.fragment, encoding), _safe_chars)
+    )
+
+def canonicalize_url(url, keep_blank_values=True, keep_fragments=False,
```

Comment:
Can this function and it's helpers live in w3lib? It would be much easier this way IMO, because it belongs to w3lib, and all it's tests are there

Reply:
So do you think it is better to keep canonicalize_url in w3lib and create some sort of helper functions for it from urlparse4?

Reply:
(btw, slack is ~down right now)

There are pros and cons in both approaches:
- if we keep it in w3lib, then it's easier to make changes to both pure python and cython version
- but if we keep it in urlparse4, then w3lib can be kept a pure-python package, and use faster version from urlparse4 if it's available.

I would keep it in urlparse4 for the moment. But we definitely need some way to run Cython version against w3lib tests.

Reply:
Thank you for explaining it! I will keep developing canonicalize_url on urlparse4 for now! 
I'm opening another PR for this :)